loki/transformations/sanitise/sequence_associations.py

```diff
@@ -0,0 +1,136 @@
+    :any:`Transformation` that resolves sequence association patterns
+    in :any:`CallStatement` nodes.
+
+    Parameters
+    ----------
+    resolve_sequence_associations : bool
+        Flag to trigger or suppress resolution of sequence associations
+    """
+
+    def __init__(self, resolve_sequence_associations=True):
+        self.resolve_sequence_associations = resolve_sequence_associations
+
+    def transform_subroutine(self, routine, **kwargs):  # pylint: disable=unused-argument
+        if self.resolve_sequence_associations:
+            do_resolve_sequence_association(routine)
+
+
+def check_if_scalar_syntax(arg, dummy):
+    """
+    Check if an array argument, arg,
+    is passed to an array dummy argument, dummy,
+    using scalar syntax. i.e. arg(1,1) -> d(m,n)
+
+    Parameters
+    ----------
+    arg:   variable
+    dummy: variable
+    """
+    if isinstance(arg, Array) and isinstance(dummy, Array):
+        if arg.dimensions:
+            if not any(isinstance(d, RangeIndex) for d in arg.dimensions):
+                return True
+    return False
+
+
+def do_resolve_sequence_association(routine):
+    """
+    Housekeeping routine to replace scalar syntax when passing arrays
+    as arguments For example, a call like
+
+    .. code-block::
+
+        real :: a(m,n)
+
+        call myroutine(a(i,j))
+
+    where myroutine looks like
+
+    .. code-block::
+
+        subroutine myroutine(a)
+            real :: a(5)
+        end subroutine myroutine
+
+    should be changed to
+
+    .. code-block::
+
+        call myroutine(a(i:m,j)
+
+    Parameters
+    ----------
+    routine : :any:`Subroutine`
+        The subroutine where calls will be changed
+
+    """
+
+    routine.body = SequenceAssociationTransformer(inplace=True).visit(routine.body)
+
+
+class SequenceAssociationTransformer(Transformer):
+    """
+    Transformer that resolves sequence association patterns in
+    :any:`CallStatement` nodes.
+    """
+
+    def visit_CallStatement(self, call, **kwargs):  # pylint: disable=unused-argument
+        """
+        Resolve sequence association patterns in arguments and return
+        new :any:`CallStatement` object if any were found.
+        """
+        if call.procedure_type is BasicType.DEFERRED:
+            return call
+
+        new_args = []
+        found_scalar = False
+        for dummy, arg in call.arg_map.items():
+            if check_if_scalar_syntax(arg, dummy):
+                found_scalar = True
+
+                n_dims = len(dummy.shape)
+                new_dims = []
+
+                if not arg.shape:
+                    # Hack: If we don't have a shape, short-circuit here
+                    new_dims = tuple(RangeIndex((None, None)) for _ in dummy.shape)
+                else:
+                    for s, lower in zip(arg.shape[:n_dims], arg.dimensions[:n_dims]):
+                        if isinstance(s, RangeIndex):
+                            new_dims += [RangeIndex((lower, s.stop))]
```

Comment:
This is probably quite a common use case, and so could the tests please be extended to cover this?

Reply:
Yes, it is and it was not covered in Santeri's original testing. I've slightly refactored his test and added specific cases for this. Please check if this now covers it.